src/skmatter/metrics/_pairwise.py

```diff
@@ -0,0 +1,213 @@
+    """
+
+    X, Y = check_pairwise_arrays(X, Y)
+
+    if X_norm_squared is not None:
+        X_norm_squared = check_array(X_norm_squared, ensure_2d=False)
+        original_shape = X_norm_squared.shape
+        if X_norm_squared.shape == (X.shape[0],):
+            X_norm_squared = X_norm_squared.reshape(-1, 1)
+        if X_norm_squared.shape == (1, X.shape[0]):
+            X_norm_squared = X_norm_squared.T
+        if X_norm_squared.shape != (X.shape[0], 1):
+            raise ValueError(
+                f"Incompatible dimensions for X of shape {X.shape} and "
+                f"X_norm_squared of shape {original_shape}."
+            )
+
+    if Y_norm_squared is not None:
+        Y_norm_squared = check_array(Y_norm_squared, ensure_2d=False)
+        original_shape = Y_norm_squared.shape
+        if Y_norm_squared.shape == (Y.shape[0],):
+            Y_norm_squared = Y_norm_squared.reshape(1, -1)
+        if Y_norm_squared.shape == (Y.shape[0], 1):
+            Y_norm_squared = Y_norm_squared.T
+        if Y_norm_squared.shape != (1, Y.shape[0]):
+            raise ValueError(
+                f"Incompatible dimensions for Y of shape {Y.shape} and "
+                f"Y_norm_squared of shape {original_shape}."
+            )
+
+    if cell is None:
+        return _euclidean_distances(X, Y, X_norm_squared, Y_norm_squared, squared)
+    else:
+        return _periodic_euclidean_distances(X, Y, squared=squared, cell=cell)
+
+
+def _periodic_euclidean_distances(X, Y=None, *, squared=False, cell=None):
+    X, Y = np.array(X).astype(float), np.array(Y).astype(float)
+    XY = np.concatenate([X - y for y in Y])
+    XY -= np.round(XY / cell) * cell
+    distance = np.linalg.norm(XY, axis=1).reshape(X.shape[0], Y.shape[0], order="F")
+    if squared:
+        distance **= 2
+    return distance
+
+
+def pairwise_mahalanobis_distances(
+    X: np.ndarray,
+    Y: np.ndarray,
+    cov_inv: np.ndarray,
+    cell: Union[np.ndarray, None] = None,
+    squared: bool = False,
+):
+    r"""
+    Calculate the pairwise Mahalanobis distance between two arrays. This metric is used
+    for calculating the distances between observations from Gaussian distributions. It
+    is defined as:
+
+    .. math::
+        d_{\Sigma}(x, y)^2 = (x - y)^T \Sigma^{-1} (x - y)
+
+    where :math:`\Sigma` is the covariance matrix, :math:`x` and :math:`y` are
+    observations from the same distribution.
+
+    Parameters:
+        x : np.ndarray of shape (n_samples_X, n_features)
+            An array where each row is a sample and each column is a feature.
+        y : np.ndarray of shape (n_samples_Y, n_features)
+            An array where each row is a sample and each column is a feature.
+        cov_inv : np.ndarray
+            The inverse covariance matrix of shape (n_features, n_features).
+        cell : np.ndarray, optinal
+            The cell size for periodic boundary conditions.
+        squared : bool
+            Whether to return the squared distance.
```

Comment:
Do we really need this. I think this is just cluttering the parameters and squaring is not that hard.

Reply:
This is inherited from sklearn. See https://github.com/scikit-learn/scikit-learn/blob/f61dd6cfc0263e4f6b8f9f223ce1bc3510328057/sklearn/metrics/pairwise.py#L364

Reply:
I see.